### PR TITLE
Fix storyboard `UseSkinSprites` being implemented incorrectly

### LINKED ISSUE #24813: Looking up skin sprites from storyboards is partially broken
### Type

Game behaviour

### Bug description

Wanted to check if https://github.com/ppy/osu/pull/24809 doesn't break looking up skin sprites, and found it was already broken on master. Was 100% working before (https://github.com/ppy/osu/pull/9786), must have broken somewhere along the way. https://github.com/ppy/osu/pull/17703 looks like a likely candidate.

Reproduction:

1. Download https://osu.ppy.sh/beatmapsets/832364#osu/1743837
2. Autoplay the map
3. Sprites that should be looked up from the skin (`hitcircleoverlay`, `hitcirclenumber`) at around 00:04:102 won't be, and won't respond to beatmap skin toggle

Not sure if the above conclusions as to where this broke are correct, but the logic looks 100% wrong:

https://github.com/ppy/osu/blob/5bda9a3024c46f8f7d142d6bb761a22a4e83c960/osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs#L93-L99

If the asset is provided as part of the beatmap skin, it will also be available to the storyboard via `GetTextureFromPath()`. Therefore the skin-querying pathway will never be activated, and things like toggling the beatmap skin toggle won't be responded to.

### Screenshots or videos

osu_2023-09-14_15-48-31

### Version

master

### Logs

n/a

### PR DESCRIPTION
**Pull request by peppy**
This was implemented as a "fallback", but it's actually intended to be an "override". As in it allows storyboarders to *prefer* a skin sprite before falling back to a local version contained within the storyboard.

Can be tested with https://osu.ppy.sh/beatmapsets/832364#osu/1743837.

Closes https://github.com/ppy/osu/issues/24813.

### CODE REVIEW

**Comment by bdach** [comment]
The sprites are scaled wrong if the skin is changed during gameplay:

https://github.com/ppy/osu/assets/20418176/d949c8f7-a133-4f45-97a8-281d8a822a15

**Comment by peppy** [comment]
How though? This is just changing the texture.

I think it's more likely related to skins with `@2x` resource vs ones without?

**Comment by bdach** [comment]
> I think it's more likely related to skins with `@2x` resource vs ones without?

`@2x` resources being present is largely irrelevant to this issue. This is more apparent on the following screenshot:

osu_2023-09-20_19-26-54

The actual root cause of this is that a `DrawableStoryboardSprite` is a `Sprite`. For convenience, `Sprite`s will use the texture's size as their own size, but _only the first time the texture is assigned to the sprite_. Subsequent swaps of the texture will change the texture, but will not adjust the size of the sprite in kind. [This is actually even documented behaviour.](https://github.com/ppy/osu-framework/blob/9ca2a05e3246d525852b395443af3782796f249b/osu.Framework/Graphics/Sprites/Sprite.cs#L143-L148)

Another factor here is that texture-based logic has 0 chance of ever working on non-legacy skins, but not sure how relevant that is and how far do we want to even be supporting this sort of mechanic...

**Comment by peppy** [comment]
I can't seem to repro this issue, could you please post a skin that it's occurring on for you?

**Comment by bdach** [comment]
Here are two where it's most prominent:

https://drive.google.com/file/d/1JnUzbclZyEosMIlJqlxTqgB6F8DrGXyZ/view?usp=drive_link
https://drive.google.com/file/d/1qpEZbmJBveMEDKtNUeZ51_RaTbMDjjtd/view?usp=sharing

Relevant sprites:

| skin 1 | skin 2 |
| :-: | :-: |
| hitcircle@2x | hitcircle@2x |
| default-1@2x | default-1@2x |

**Comment by peppy** [comment]
> Here are two where it's most prominent

Thanks, was able to reproduce.

**Comment by peppy** [comment]
I've pushed a simple fix for this one. Doesn't look like it affected `DrawableStoryboardAnimations` as `TextureAnimation` already has [local sizing logic](https://github.com/ppy/osu-framework/blob/master/osu.Framework/Graphics/Animations/TextureAnimation.cs#L37-L38).

File: osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs [hunk 1]
```diff
@@ -84,24 +90,28 @@ public DrawableStoryboardSprite(StoryboardSprite sprite)
...
 
             Sprite.ApplyTransforms(this);
         }
 
-        private void skinSourceChanged() => Texture = skin.GetTexture(Sprite.Path);
+        private void skinSourceChanged()
+        {
+            // Setting texture will only update the size if it's zero.
+            // So let's force an update by setting to zero.
+            Size = Vector2.Zero;
```
Comment (bdach) [inline osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs:111]: This will fail [if this early guard](https://github.com/ppy/osu-framework/blob/ff3829952c3a39b7bcf637b531d840f3df6ab2a9/osu.Framework/Graphics/Sprites/Sprite.cs#L154-L155) is hit, which I've hit in practice.

Suggest applying

```diff
diff --git a/osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs b/osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs
index 87c09c3e5e..68892a6fc1 100644
--- a/osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs
+++ b/osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs
@@ -106,11 +106,11 @@ private void load(Storyboard storyboard)
 
         private void skinSourceChanged()
         {
-            // Setting texture will only update the size if it's zero.
-            // So let's force an update by setting to zero.
-            Size = Vector2.Zero;
-
             Texture = skin.GetTexture(Sprite.Path) ?? textureStore.Get(Sprite.Path);
+
+            // Setting texture will only update the size if it's zero.
+            // So let's force an explicit update.
+            Size = new Vector2(Texture.DisplayWidth, Texture.DisplayHeight);
         }
 
         protected override void Dispose(bool isDisposing)

```

instead.

Comment (peppy) [reply osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs:111]: Is what I originally had. Sounds fine.

**Comment by bdach** [comment]
Looking at my random collection of skins, this also fails to work correctly on some "special" skins, which for instance have `HitCirclePrefix` set in `skin.ini`. One example is https://drive.google.com/file/d/13l5KKJHtsTlmNrxBooiSi4CO8E_qVF2g/view?usp=sharing.

I don't know how you fix that one without using skinnables in storyboards to share the ridiculous lookup logic. Or we just wontfix it, I dunno.

**Comment by peppy** [comment]
It looks like stable fails in the same way, so let's ignore that edge case 🙈 .

**Review by bdach — DISMISSED** [review]

**Comment by bdach** [comment]
Fixed an issue I didn't think of caught by the tests (null texture case). Will let this run past CI first just to be safe.

**Review by bdach — APPROVED** [review]
